File: backend/tools/testRedisConnection.py
#!/usr/bin/env python3
import os
import sys
import redis
import dotenv
import ssl
from pathlib import Path

# Load environment variables
dotenv.load_dotenv(Path(__file__).parent.parent.parent / '.env')

def test_redis_connection():
    redis_password = os.environ.get('REDIS_PASSWORD')
    use_tls = os.environ.get('REDIS_SSL') == 'true'
    
    if not redis_password:
        print('REDIS_PASSWORD environment variable is required')
        sys.exit(1)
    
    # Configure SSL options for Redis client
    def get_ssl_context():
        if not use_tls:
            return None
        
        try:
            cert_path = Path(__file__).parent.parent.parent / 'certs/redis.crt'
            key_path = Path(__file__).parent.parent.parent / 'certs/redis.key'
            ca_path = Path(__file__).parent.parent.parent / 'certs/server.crt'
            
            if cert_path.exists() and key_path.exists() and ca_path.exists():
                ssl_context = ssl.create_default_context()
                ssl_context.load_cert_chain(cert_path, keyfile=key_path)
                ssl_context.load_verify_locations(ca_path)
                ssl_context.check_hostname = False
                ssl_context.verify_mode = ssl.CERT_NONE  
                return ssl_context
            else:
                print('SSL certificates not found, using basic SSL')
                ssl_context = ssl.create_default_context()
                ssl_context.check_hostname = False
                ssl_context.verify_mode = ssl.CERT_NONE
                return ssl_context
        except Exception as e:
            print(f'Error loading Redis SSL certificates: {str(e)}')
            ssl_context = ssl.create_default_context()
            ssl_context.check_hostname = False
            ssl_context.verify_mode = ssl.CERT_NONE
            return ssl_context
    
    try:
        print('Connecting to Redis...')
        print(f'TLS enabled: {use_tls}')
        
        # Create Redis client
        redis_client = redis.Redis(
            host='localhost',
            port=6379,
            password=redis_password,
            ssl=use_tls,
            ssl_cert_reqs=None if use_tls else None,
            ssl_keyfile=str(Path(__file__).parent.parent.parent / 'certs/redis.key') if use_tls else None,
            ssl_certfile=str(Path(__file__).parent.parent.parent / 'certs/redis.crt') if use_tls else None,
            ssl_ca_certs=str(Path(__file__).parent.parent.parent / 'certs/server.crt') if use_tls else None,
            decode_responses=True
        )
        
        # Test setting and getting a value
        test_key = f'test_key_{os.urandom(4).hex()}'
        test_value = f'Connection successful - {str(datetime.datetime.now())}'
        
        redis_client.set(test_key, test_value)
        retrieved_value = redis_client.get(test_key)
        
        print(f'Test value retrieved: {retrieved_value}')
        
        # Clean up
        redis_client.delete(test_key)
        print('Redis connection test completed successfully')
        
    except Exception as e:
        print(f'Failed to connect to Redis: {str(e)}')
        sys.exit(1)

if __name__ == '__main__':
    test_redis_connection()

File: backend/tools/testRedisConnection.js
#!/usr/bin/env node
// backend/tools/testRedisTLS.js
const Redis = require('redis');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Helper function to verify TLS is being used
async function verifyTlsConnection(client) {
  // Generate a random key
  const testKey = `tls-test-${crypto.randomBytes(8).toString('hex')}`;
  const testValue = `value-${Date.now()}`;
  
  try {
    // Set a value
    await client.set(testKey, testValue);
    console.log('✅ Successfully wrote data to Redis');
    
    // Get the value back
    const retrievedValue = await client.get(testKey);
    console.log('✅ Successfully read data from Redis:', retrievedValue);
    
    // Verify the value matches
    if (retrievedValue === testValue) {
      console.log('✅ Data integrity verified');
      
      // Clean up the test key
      await client.del(testKey);
      console.log('✅ Successfully deleted test key');
      
      return true;
    } else {
      console.error('❌ Data integrity check failed');
      return false;
    }
  } catch (error) {
    console.error('❌ TLS verification operation failed:', error);
    return false;
  }
}

async function testRedisTLS() {
  console.log('Redis TLS Connection Test Tool');
  console.log('===============================');
  
  // Load environment variables
  const redisPassword = process.env.REDIS_PASSWORD;
  if (!redisPassword) {
    console.error('Redis password not found in environment variables');
    process.exit(1);
  }
  
  try {
    // Find SSL certificates - prioritize container paths
    let certPath, keyPath, caPath;
    const containerCertPath = path.join(__dirname, '../certs/redis.crt');
    const containerKeyPath = path.join(__dirname, '../certs/redis.key');
    const containerCaPath = path.join(__dirname, '../certs/server.crt');
    
    if (fs.existsSync(containerCertPath) && fs.existsSync(containerKeyPath)) {
      console.log('Found certificates in container path');
      certPath = containerCertPath;
      keyPath = containerKeyPath;
      caPath = containerCaPath;
    } else {
      // Fall back to project root
      const rootCertPath = path.join(__dirname, '../../certs/redis.crt');
      const rootKeyPath = path.join(__dirname, '../../certs/redis.key');
      const rootCaPath = path.join(__dirname, '../../certs/server.crt');
      
      if (fs.existsSync(rootCertPath) && fs.existsSync(rootKeyPath)) {
        console.log('Found certificates in project root');
        certPath = rootCertPath;
        keyPath = rootKeyPath;
        caPath = rootCaPath;
      } else {
        console.error('SSL certificates not found');
        process.exit(1);
      }
    }
    
    console.log('Creating Redis client with TLS configuration');
    const client = Redis.createClient({
      socket: {
        host: 'localhost', // Use redis for Docker connections
        port: 6379,
        tls: {
          cert: fs.readFileSync(certPath),
          key: fs.readFileSync(keyPath),
          ca: fs.readFileSync(caPath),
          rejectUnauthorized: false,
          minVersion: 'TLSv1.2',
          maxVersion: 'TLSv1.3'
        }
      },
      password: redisPassword
    });
    
    // Register event handlers
    client.on('error', (err) => {
      console.error('Redis client error:', err);
    });
    
    client.on('connect', () => {
      console.log('✅ Redis client connected');
    });
    
    client.on('ready', async () => {
      console.log('✅ Redis client ready');
      const verified = await verifyTlsConnection(client);
      
      if (verified) {
        console.log('===============================');
        console.log('✅ Redis TLS connection is working properly!');
        console.log('===============================');
      } else {
        console.log('===============================');
        console.error('❌ Redis TLS connection verification failed!');
        console.log('===============================');
      }
      
      await client.quit();
      console.log('Redis client disconnected');
    });
    
    // Connect to Redis
    console.log('Connecting to Redis...');
    await client.connect();
  } catch (error) {
    console.error('Fatal error during Redis TLS test:', error);
    process.exit(1);
  }
}

// Run the test
testRedisTLS().catch(error => {
  console.error('Unhandled error:', error);
  process.exit(1);
});